### ISSUE TITLE
Attribute for type on VRFrameOfReference

### ISSUE THREAD

**Issue by toji** [w3c/webvr#275]
I'm wondering if it wouldn't be a good idea to have a `type` attribute on `VRFrameOfReference`. Doing so would enable apps to track a little less state and enable introspection of the Frame of Reference objects that's not possible without application cooperation otherwise.

For example: If an application is trying to use a "stage" FoR but falls back to an "eyeLevel" FoR if it's not available, the app will need to do a height translation of the view matrix before rendering, otherwise the user will appear stuck in the floor. Without a `type` attribute the code looks like this:

```js
session.requestFrameOfReference("stage").then((frameOfRef) => {
  vrFrameOfRef = frameOfRef;
  adjustHeight = false;
}).catch(() => {
  session.requestFrameOfReference("eyeLevel").then((frameOfRef) => {
    vrFrameOfRef = frameOfRef;
    adjustHeight = true;
  });
});

function onVRFrame(frame) {
  let pose = frame.getDevicePose(vrFrameOfRef);
  for (let view of frame.views) {
    let viewMat = pose.getViewMatrix(view);
    if (adjustHeight) {
      vewMatrix = translateMatrixY(viewMatrix, -PLAYER_HEIGHT);
    }
    // Render with view matrix
  }
};
```

With the attribute it can be:

```js
session.requestFrameOfReference("stage").then((frameOfRef) => {
  vrFrameOfRef = frameOfRef;
}).catch(() => {
  session.requestFrameOfReference("eyeLevel").then((frameOfRef) => {
    vrFrameOfRef = frameOfRef;
  });
});

function onVRFrame(frame) {
  let pose = frame.getDevicePose(vrFrameOfRef);
  for (let view of frame.views) {
    let viewMat = pose.getViewMatrix(view);
    if (vrFrameOfRef.type != "stage") {
      vewMatrix = translateMatrixY(viewMatrix, -PLAYER_HEIGHT);
    }
    // Render with view matrix
  }
};
```

That's a minor difference, but it does allow applications to simplify the amount of external state they're tracking and can make situations where multiple sessions using multiple frames of reference are being tracked simultaneously much simpler and less error prone. (This will commonly be the case for a page that supports both magic window and exclusive presentation.)

It also means that tooling can quickly identify the type of `VRFrameOfReference` being used, which can't be determined by normal JS RTTI methods since they all return the same object type.

Thoughts?

**Comment by huningxin** [comment]
+1. It looks good to me.

**Comment by TrevorFSmith** [comment]
Having a type attribute would allow the `session.requestFrameOfReference` call to accept a list of acceptable types in order of preference, which would prevent the need for nested promise handling:

`
session.requestFrameOfReference("stage", "eyeLevel").then((frameOfRef) => {
  // frameOfRef.type determines whether it is stage or eyeLevel
}).catch(() => {
  // no usable FrameOfReference was found
});
`

**Comment by toji** [comment]
@TrevorFSmith: I think part of your particular scenario may be addressed by #283?

**Comment by TrevorFSmith** [comment]
I'm looking forward a bit to when we have additional frames of reference, for example some sort of geospatial frame of reference for AR. An app that is made to work with a variety of FoRs could ask for the types "geospatial", "stage", "eyeLevel" and the system could return the first supported type, which the app could check using the proposed `type` attribute.

**Comment by NellWaliczek** [comment]
Maybe I'm underthinking this, but couldn't the code simply have the PLAYER_HEIGHT offset be set to zero when using the stage and a non-zero value otherwise?  Or are you trying to save time on the one math operation?

**Comment by toji** [comment]
That particular scenario is effectively addressed by #283, so the only outstanding item I see in this issue is Trevor's request for accepting a list of possible types. I'm not against the idea, really, but I think that it's not a problem with the current set of FoRs that we have now and we should probably defer adding that feature until it has a concrete use. Going to close this now.
